### DISCUSSION TITLE
bug: Sse calling it self even after completion

### DISCUSSION THREAD

**Discussion by JAYBORICHA07** [trpc/trpc#5840] [Q&A]
### Provide environment information

I am using trpc with reactQuery, fastify as backend and React.js as frontend.

here is my backend code : 

```
import { protectedProcedure, router } from "../context.trpc";
import { observable } from "@trpc/server/observable";
import z from "zod";

export const newLinkAuditController = router({
  assessment: protectedProcedure
    .input(z.object({ count: z.number() }))
    .subscription(({ input }) =>
      observable<{ msg: number }>((sub) => {
        let i = 0;
        const jut = setInterval(() => {
          if (i === input.count) {
            console.log("completed");
            sub.complete();
            clearInterval(jut);
          } else {
            i++;
            sub.next({ msg: i });
          }
        }, 1000);
      })
    ),
});
```

here is my frontend-code
```
const handleFormSubmit = async ({ url }: { url: string }) => {
    setIsLoading(true);
    const abc = trpcFetch.newLinkAudit.subscribe(
      {
        count: 9,
      },
      {
        onError: (err) => {
          console.log(err);
        },
        onComplete: () => {
          console.log("Done!!");
        },
        onData: (data) => {
          console.log(data);
        },
      }
    );
    return;
  };
```

### Describe the bug

At start in my handleformsubmit i am using sse and it is calling it self again and again i am receiving data in the onData part in frontend but not able to console Done!! in onSubmit. it is calling it self again.

### Link to reproduction

sorry there is no link i have uploaded the code above

### To reproduce

i have provided the fastify backend and react frontend code above

### Additional information

This was the problem

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5837">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5837/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/trpc/trpc/issues/5837/pledge.svg">
</picture>
</a>

**Comment by KATT** [comment]
The subscription will be continued until you `.unsubscribe()`. This doesn't look like a bug. Converting to discussion

**Reply by JAYBORICHA07** [reply]
Oh can you please tell me where should I user this .unsubscribe method 

And also thanks for the cleaning that.

**Reply by krishna-404** [reply]
Oh! I would also love to understand this. Facing the same issue.

**Reply by KATT** [reply]
`abc.unsubscribe()` when you have the result you want

Or use the React hook that does it automatically when unmounting. You can use `skipToken` to call it conditionally

**Reply by KATT** [reply]
Here's an example of a subscription that uses `skipToken` in a hook

https://github.com/trpc/trpc/blob/48de686d22ccbefe1cf97eaa7f205dabe7302d2a/examples/next-sse-chat/src/app/channels/%5BchannelId%5D/hooks.ts#L107-L118

**Reply by JAYBORICHA07** [reply]
Hello @KATT,  I changed the code and it's working as desired behaviour but  I am having some other problems too that whenever 
 I send error from the backend with sub.error("Sample error")  I am not receiving it on the frontend in the onError section i am attaching my code snippets below.
 
 
 Frontend
 ```
 const abc = trpcFetch.newLinkAudit.subscribe(
      { url },
      {
        onComplete: () => {
          message.success("Landing page audit completed successfully");
        },
        onData: (data) => {
          setLinksRecord((prev) => [...prev, data]);
          setIsSubmitted(true);
          if (data.key === "8") {
            abc.unsubscribe();
            setIsLoading(false);
          }
        },
        onError: (error) => {
          console.error("Subscription error:", error);
          message.error(`Error during audit: ${error.message}`);
          setIsSubmitted(false);
          abc.unsubscribe();
        },
      }
    );
 ```
 
 Backend
 ```
 export const newLinkAuditController = router({
  assessment: protectedProcedure
    .input(z.object({ url: z.string().url() }))
    .subscription(async ({ ctx, input }) =>
      observable<{
        key: string;
        title: string;
        description: string;
        assessment?: string;
        improvement?: string;
        layoutAndStructure?: string;
        layoutAndStructureImprovement?: string;
        designAndAesthetics?: string;
        designAndAestheticsImprovement?: string;
        usabilityAndAccessibility?: string;
        usabilityAndAccessibilityImprovement?: string;
        contentPresentation?: string;
        contentPresentationImprovement?: string;
        visualoverallscore?: string;
        score?: string;
      }>(async (sub) => {
        try {
          // some API calls to ChatGPT
          const ProblemResponse = await ContentAssessCriteria("problem");

          sub.next({
            key: "1",
            title: "Problem",
            description: descriptions["problem"],
            assessment: ProblemResponse.problem.assessment,
            improvement: ProblemResponse.problem.improvement,
            score: `${ProblemResponse.problem.score}`,
          });
          return;
        } catch (error) {
          if (error instanceof OpenAiError) {
            switch (error.code) {
              case "INVALID_API_KEY":
                sub.error(
                  "Invalid OpenAI API key. Please check your credentials."
                );

              case "RATE_LIMIT_EXCEEDED":
                sub.error(
                  "OpenAI API rate limit exceeded. Please try again later."
                );
              case "CREDENTIALS_EXHAUSTED":
                sub.error(
                  "OpenAI API credentials exhausted or insufficient credit. Please check your account."
                );
              default:
                sub.error(error.message);
            }
            return;
          } else if (error instanceof SyntaxError) {
            sub.error("Error parsing OpenAI response");
            return;
          } else {
            sub.error(error.message || "An unexpected error occurred");
            return;
          }
        }
      })
    ),
});
 ``` 
 
with this logic whenever some error occurs the server is getting crashed and i am not able to receive the error on the frontend.

**Reply by krishna-404** [reply]
Thanks for this :) 

Why doesnt `sub.complete` on the backend work?

**Reply by krishna-404** [reply]
Hi! Just wanted to let you know that I would be happy to contribute if this is a bug or update the docs for being more verbose here. Thanks :)

**Comment by pg-wtatum** [comment]
I agree that it would be helpful for trfc itself to have a way of handling an expected end of a stream. The server side constructs have the right way to signal this (returning from generator or calling complete on observable) but it seems like the client side doesn't make any distinction between an unexpected end of the EventSource and the "completion" signal .

This feels out of line with RxJS, reactivestreams, and other APIs that allow developers to clearly indicate that a stream is over and will not produce any more data.

Obviously we could capture this "in the stream" by having a different object for "a value" and "no more values after this" -- but that feels like a lot of ceremony that the framework could possibly handle?
